### ISSUE TITLE
manifest processing model, what if null base URL? (related to origin issue)

### ISSUE THREAD

**Issue by danielweck** [w3c/wpub#374]
Issue originally raised in the "opaque origin" conversation:
https://github.com/w3c/wpub/issues/321#issuecomment-444409200

**Comment by iherman** [comment]
If the manifest is embedded, the only way this can happen (see https://github.com/w3c/wpub/issues/321#issuecomment-444352620) is if the value of `baseURI` in the DOM for the `<script>` element is null. The question is when would that happen per the HTML or the DOM specs. I do not have a precise answer, but I _suspect_ that it may happen in the case of a `file:` URL, ie, when the entry page is read from the file system. If, as we referred to in #321, we disallow that (or we just say the effect depends on user agent and users should be prepared) then we are done, aren't we?

**Comment by iherman** [comment]
One step further in https://www.w3.org/TR/DOM-Level-3-Core/core.html#Node3-baseURI:

> **baseURI of type DOMString, readonly, introduced in DOM Level 3**
> The absolute base URI of this node or null if the implementation wasn't able to obtain an absolute URI.

**Comment by iherman** [comment]
Related to the original question: I am fine modifying the processing model stating that if this happens, the processing stops.

**Comment by danielweck** [comment]
I my original comment I mentioned `data:` URLs. I believe this is a problematic edge case.
https://github.com/w3c/wpub/issues/321#issuecomment-444203330

**Comment by danielweck** [comment]
In the following edge case example, the `data:` URL encodes an HTML document which does not specify `<base href="...">` in its `head`. Consequently, the `<script>`-embedded WebPub manifest has a null base URI (inherited from its parent document context).

Please ignore the lack of character escaping, this is pseudo-code:

`https://domain.org/index.html`
=>
```
<html>
<body>
<iframe
    src="data:text/html,<html><head><script type="application/ld+json">{...}</script></head><body>...</body></html>"
/>
</body>
</html>
```

Let's not try to explain why such convoluted markup would exist in the first place. Let's just handle the edge case regardless of its possible causes. I see two options:
1) Early termination: no point continuing to load the WebPub manifest without a base URI. If I understand correctly, at this point in time the JSON-LD processing model is being discussed / finalized, with respect to handling base URI in embedded contexts. However the WebPub processing model can isolate itself from this potentially moving target, by aborting as soon as the failure criterion is met.
2) Allow the WebPub manifest to load: if/when a base URI is required as part of the JSON-LD processing model in order to resolve an absolute URL from a relative "path", and this base URI is missing, then let the JSON-LD processor raise the appropriate error. This may be a complete abort, of a skip-resource-and-continue kind of algorithm (I am not sure, do you know Ivan?)

**Comment by iherman** [comment]
(2) is of course sounds as a viable and reasonable option, except that I would expect many reading systems would want to parse and interpret the manifest directly for the purposes of publications without relying on a full-blown json-ld processor. I.e., relying on that may be an issue.

On (1) yes, there are discussions on the JSON-LD but on (other) edge cases of embedding a manifest (e.g., is it required to escape certain HTML terms within the script element). I actually do not think this type of edge case has been discussed or not. Yes, the WebPub model can isolate itself, but I would think it is better to align with the JSON-LD WG. 

Bottom line, I think this question should be raised in the JSON-LD WG. I can of course raise the issue, but it may be better if you did it (on https://github.com/w3c/json-ld-syntax/issues).

Do you know what will the baseURI value be on the DOM element for <script>? Will it be null (which I expect to be)?

**Comment by danielweck** [comment]
Quick test:

```html
<html>
<body>
<iframe
    width="100%"
    height="100%"

    src="data:text/html;base64,CjxodG1sPgo8aGVhZD4KPGJhc2UgaHJlZj0iaHR0cHM6Ly9kb21haW4ub3JnL3BhdGgvIiAvPgoKPHNjcmlwdCBpZD0ic2NyaXB0IiB0eXBlPSJ0ZXh0L2phdmFzY3JpcHQiPgogIGRvY3VtZW50LmFkZEV2ZW50TGlzdGVuZXIoIkRPTUNvbnRlbnRMb2FkZWQiLCBmdW5jdGlvbihldmVudCkgewogICAgY29uc29sZS5sb2coIkRPTUNvbnRlbnRMb2FkZWQiKTsKICAgIAogICAgLy8gd2luZG93LmxvY2F0aW9uLm9yaWdpbiB0b28KICAgIGxldCB0MSA9ICJ3aW5kb3cub3JpZ2luOiAiICsgd2luZG93Lm9yaWdpbjsKICAgIGNvbnNvbGUubG9nKHQxKTsKICAgIGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCJfMSIpLmlubmVySFRNTCA9IHQxOwogICAgCiAgICBsZXQgdDIgPSAiZG9jdW1lbnQuYmFzZVVSSTogIiArIGRvY3VtZW50LmJhc2VVUkk7CiAgICBjb25zb2xlLmxvZyh0Mik7CiAgICBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgiXzIiKS5pbm5lckhUTUwgPSB0MjsKCiAgICBsZXQgdDMgPSAibG9jYXRpb24uaHJlZjogIiArIGxvY2F0aW9uLmhyZWY7CiAgICBjb25zb2xlLmxvZyh0Myk7CiAgICBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgiXzMiKS5pbm5lckhUTUwgPSB0MzsKCiAgICBsZXQgdDQgPSAic2NyaXB0LmJhc2VVUkk6ICIgKyBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgic2NyaXB0IikuYmFzZVVSSTsKICAgIGNvbnNvbGUubG9nKHQ0KTsKICAgIGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCJfNCIpLmlubmVySFRNTCA9IHQ0OwogIH0pOwo8L3NjcmlwdD4KPC9oZWFkPgo8Ym9keT4KPGgxIGlkPSJfMSI+MTwvaDE+CjxoMSBpZD0iXzMiPjM8L2gxPgo8aDEgaWQ9Il8yIj4yPC9oMT4KPGgxIGlkPSJfNCI+NDwvaDE+CjwvYm9keT4KPC9odG1sPg=="
/>
</body>
</html>
<!--
<html>
<head>
<base href="https://domain.org/path/" />

<script id="script" type="text/javascript">
  document.addEventListener("DOMContentLoaded", function(event) {
    console.log("DOMContentLoaded");
    
    // window.location.origin too
    let t1 = "window.origin: " + window.origin;
    console.log(t1);
    document.getElementById("_1").innerHTML = t1;
    
    let t2 = "document.baseURI: " + document.baseURI;
    console.log(t2);
    document.getElementById("_2").innerHTML = t2;

    let t3 = "location.href: " + location.href;
    console.log(t3);
    document.getElementById("_3").innerHTML = t3;

    let t4 = "script.baseURI: " + document.getElementById("script").baseURI;
    console.log(t4);
    document.getElementById("_4").innerHTML = t4;
  });
</script>
</head>
<body>
<h1 id="_1">1</h1>
<h1 id="_3">3</h1>
<h1 id="_2">2</h1>
<h1 id="_4">4</h1>
</body>
</html>
-->
```

Result:

```
window.origin: null

location.href: data:text/html;base64,LONG_BASE64_STRING

document.baseURI: https://domain.org/path/

script.baseURI: https://domain.org/path/
```

If the `<base href="https://domain.org/path/" />` element is removed, then `baseURI` for both `document` and `script` is in fact not null, it is the same as `location.href` (i.e. the `data:` URL) ... which cannot be used for resolving absolute URLs from relative paths anywhere in the document (such as when processing an embedded WebPub manifest).

Based on this simple experiment, I am starting to wonder whether ; just like opaque `origin` ; the WebPub specification should simply remain silent about `baseURI` edge cases. Once again, I think that the rationale for explicitly null-testing `origin/baseURI` (e.g. fail => terminate) in the WP manifest acquisition algorithm should be that `origin/baseURI` is explicitly needed later in the algorithm. For `origin`, the processing steps rely on the fetch API response status (e.g. bad CORS -> error response). For `baseURI`, it depends on whether the WebPub specification describes in great detail how to resolve absolute URLs in the manifest, or if this is "automatically" inherited from the JSON-LD processing model (in which case this concern becomes an polyfill / user-agent implementation detail).

Thoughts?

**Comment by iherman** [comment]
Taking this out from @danielweck's long comment for an easier reference: 

> Based on this simple experiment, I am starting to wonder whether ; just like opaque origin ; the WebPub specification should simply remain silent about baseURI edge cases. Once again, I think that the rationale for explicitly null-testing origin/baseURI (e.g. fail => terminate) in the WP manifest acquisition algorithm should be that origin/baseURI is explicitly needed later in the algorithm. For origin, the processing steps rely on the fetch API response status (e.g. bad CORS -> error response). For baseURI, it depends on whether the WebPub specification describes in great detail how to resolve absolute URLs in the manifest, or if this is "automatically" inherited from the JSON-LD processing model (in which case this concern becomes an polyfill / user-agent implementation detail).

I got to a similar conclusion, so I wholeheartedly agree. Although weird, the example with `data:` URL makes sense but, also, it is perfectly possible to create a manifest using absolute URL-s only and, consequently, the interpretation of the manifest could be oblivious to the null baseURI value.

I think for both this issue and #321 we should try to find a blanket formulation in the processing which says that if a processing step runs into an error (or a OWP related error?), then the processing would stop and there would be no manifest. (We _could_ put there an note giving examples for such situations, and we can refer to the origin or the baseURI null problem, but that should only be an informal note.) I am not sure how exactly to formulate that, but maybe @mattgarrish can come with the best terminology...

**Comment by iherman** [comment]
N.B. I have raised an explicit issue by the JSON-LD WG (https://github.com/w3c/json-ld-syntax/issues/103), a.k.a. passing over the buck:-)

**Comment by danielweck** [comment]
Thanks Ivan!

Let me also clarify this statement:

>  For baseURI, it depends on whether the WebPub specification describes in great detail how to resolve absolute URLs in the manifest, or if this is "automatically" inherited from the JSON-LD processing model (in which case this concern becomes an polyfill / user-agent implementation detail).

If the former (i.e. the WP specification describes "parsing" rules, probably as an extension to the JSON-LD processing model), then the manifest algorithm must be clear about what happens when an absolute URL cannot be resolved:
1) complete failure (i.e. abort loading the manifest entirely)
or:
2) skip the unresolved URL (i.e. ignore the resource), and continue loading the rest of the data.
